File: TESTING/EIG/dbdt04.f
      SUBROUTINE DBDT04( UPLO, N, D, E, S, NS, U, LDU, VT, LDVT, WORK, RESID );

*  -- LAPACK test routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      String             UPLO;
      int                LDU, LDVT, N, NS;
      double             RESID;
      // ..
      // .. Array Arguments ..
      double             D( * ), E( * ), S( * ), U( LDU, * ), VT( LDVT, * ), WORK( * );
      // ..

* ======================================================================

      // .. Parameters ..
      double             ZERO, ONE;
      const              ZERO = 0.0, ONE = 1.0 ;
      // ..
      // .. Local Scalars ..
      int                I, J, K;
      double             BNORM, EPS;
      // ..
      // .. External Functions ..
      bool               LSAME;
      int                IDAMAX;
      double             DASUM, DLAMCH;
      // EXTERNAL LSAME, IDAMAX, DASUM, DLAMCH
      // ..
      // .. External Subroutines ..
      // EXTERNAL DGEMM
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC ABS, DBLE, MAX, MIN
      // ..
      // .. Executable Statements ..

      // Quick return if possible.

      RESID = ZERO;
      if (N <= 0 || NS <= 0) RETURN;

      EPS = DLAMCH( 'Precision' );

      // Compute S - U' * B * V.

      BNORM = ZERO;

      if ( LSAME( UPLO, 'U' ) ) {

         // B is upper bidiagonal.

         K = 0;
         for (I = 1; I <= NS; I++) { // 20
            for (J = 1; J <= N-1; J++) { // 10
               K = K + 1;
               WORK( K ) = D( J )*VT( I, J ) + E( J )*VT( I, J+1 );
            } // 10
            K = K + 1;
            WORK( K ) = D( N )*VT( I, N );
         } // 20
         BNORM = ABS( D( 1 ) );
         for (I = 2; I <= N; I++) { // 30
            BNORM = MAX( BNORM, ABS( D( I ) )+ABS( E( I-1 ) ) );
         } // 30
      } else {

         // B is lower bidiagonal.

         K = 0;
         for (I = 1; I <= NS; I++) { // 50
            K = K + 1;
            WORK( K ) = D( 1 )*VT( I, 1 );
            for (J = 1; J <= N-1; J++) { // 40
               K = K + 1;
               WORK( K ) = E( J )*VT( I, J ) + D( J+1 )*VT( I, J+1 );
            } // 40
         } // 50
         BNORM = ABS( D( N ) );
         for (I = 1; I <= N-1; I++) { // 60
            BNORM = MAX( BNORM, ABS( D( I ) )+ABS( E( I ) ) );
         } // 60
      }

      dgemm('T', 'N', NS, NS, N, -ONE, U, LDU, WORK( 1 ), N, ZERO, WORK( 1+N*NS ), NS );

      // norm(S - U' * B * V)

      K = N*NS;
      for (I = 1; I <= NS; I++) { // 70
         WORK( K+I ) =  WORK( K+I ) + S( I );
         RESID = MAX( RESID, DASUM( NS, WORK( K+1 ), 1 ) );
         K = K + NS;
      } // 70

      if ( BNORM <= ZERO ) {
         if (RESID != ZERO) RESID = ONE / EPS;
      } else {
         if ( BNORM >= RESID ) {
            RESID = ( RESID / BNORM ) / ( DBLE( N )*EPS );
         } else {
            if ( BNORM < ONE ) {
               RESID = ( MIN( RESID, DBLE( N )*BNORM ) / BNORM ) / ( DBLE( N )*EPS );
            } else {
               RESID = MIN( RESID / BNORM, DBLE( N ) ) / ( DBLE( N )*EPS );
            }
         }
      }

      RETURN;

      // End of DBDT04

      }

File: test/eig/dbdt04.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:dart_lapack/src/blas/dasum.dart';
import 'package:dart_lapack/src/blas/dgemm.dart';
import 'package:dart_lapack/src/install/lsame.dart';
import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/install/dlamch.dart';
import 'package:dart_lapack/src/matrix.dart';

void dbdt04(
  final String UPLO,
  final int N,
  final Array<double> D_,
  final Array<double> E_,
  final Array<double> S_,
  final int NS,
  final Matrix<double> U_,
  final int LDU,
  final Matrix<double> VT_,
  final int LDVT,
  final Array<double> WORK_,
  final Box<double> RESID,
) {
  final D = D_.having();
  final E = E_.having();
  final S = S_.having();
  final U = U_.having(ld: LDU);
  final VT = VT_.having(ld: LDVT);
  final WORK = WORK_.having();
  const ZERO = 0.0, ONE = 1.0;
  int I, J, K;
  double BNORM, EPS;

  // Quick return if possible.

  RESID.value = ZERO;
  if (N <= 0 || NS <= 0) return;

  EPS = dlamch('Precision');

  // Compute S - U' * B * V.

  BNORM = ZERO;

  if (lsame(UPLO, 'U')) {
    // B is upper bidiagonal.

    K = 0;
    for (I = 1; I <= NS; I++) {
      for (J = 1; J <= N - 1; J++) {
        K++;
        WORK[K] = D[J] * VT[I][J] + E[J] * VT[I][J + 1];
      }
      K++;
      WORK[K] = D[N] * VT[I][N];
    }
    BNORM = D[1].abs();
    for (I = 2; I <= N; I++) {
      BNORM = max(BNORM, D[I].abs() + E[I - 1].abs());
    }
  } else {
    // B is lower bidiagonal.

    K = 0;
    for (I = 1; I <= NS; I++) {
      K++;
      WORK[K] = D[1] * VT[I][1];
      for (J = 1; J <= N - 1; J++) {
        K++;
        WORK[K] = E[J] * VT[I][J] + D[J + 1] * VT[I][J + 1];
      }
    }
    BNORM = D[N].abs();
    for (I = 1; I <= N - 1; I++) {
      BNORM = max(BNORM, D[I].abs() + E[I].abs());
    }
  }

  dgemm('T', 'N', NS, NS, N, -ONE, U, LDU, WORK(1).asMatrix(N), N, ZERO,
      WORK(1 + N * NS).asMatrix(NS), NS);

  // norm(S - U' * B * V)

  K = N * NS;
  for (I = 1; I <= NS; I++) {
    WORK[K + I] += S[I];
    RESID.value = max(RESID.value, dasum(NS, WORK(K + 1), 1));
    K += NS;
  }

  if (BNORM <= ZERO) {
    if (RESID.value != ZERO) RESID.value = ONE / EPS;
  } else {
    if (BNORM >= RESID.value) {
      RESID.value = (RESID.value / BNORM) / (N * EPS);
    } else {
      if (BNORM < ONE) {
        RESID.value = (min(RESID.value, N * BNORM) / BNORM) / (N * EPS);
      } else {
        RESID.value = min(RESID.value / BNORM, N) / (N * EPS);
      }
    }
  }
}
